src/Hosting/SystemdServices/src/SystemdServiceLifetime.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Extensions.Hosting.SystemdServices
+{
+    public class SystemdServiceLifetime : IHostLifetime, IDisposable
+    {
+        private readonly ManualResetEvent _shutdownBlock = new ManualResetEvent(false);
+        private CancellationTokenRegistration _applicationStartedRegistration;
+        private CancellationTokenRegistration _applicationStoppingRegistration;
+
+        public SystemdServiceLifetime(IHostEnvironment environment, IHostApplicationLifetime applicationLifetime, ISystemdNotifier systemdNotifier, ILoggerFactory loggerFactory)
+        {
+            Environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            ApplicationLifetime = applicationLifetime ?? throw new ArgumentNullException(nameof(applicationLifetime));
+            SystemdNotifier = systemdNotifier ?? throw new ArgumentNullException(nameof(systemdNotifier));
+            Logger = loggerFactory.CreateLogger("Microsoft.Hosting.Lifetime");
+        }
+
+        private IHostApplicationLifetime ApplicationLifetime { get; }
+        private IHostEnvironment Environment { get; }
+        private ILogger Logger { get; }
+        private ISystemdNotifier SystemdNotifier { get; }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            return Task.CompletedTask;
+        }
+
+        public Task WaitForStartAsync(CancellationToken cancellationToken)
+        {
+            _applicationStartedRegistration = ApplicationLifetime.ApplicationStarted.Register(state =>
+            {
+                ((SystemdServiceLifetime)state).OnApplicationStarted();
+            },
+            this);
+            _applicationStoppingRegistration = ApplicationLifetime.ApplicationStopping.Register(state =>
+            {
+                ((SystemdServiceLifetime)state).OnApplicationStopping();
+            },
+            this);
+
+            // systemd sends SIGTERM to stop the service.
+            AppDomain.CurrentDomain.ProcessExit += OnProcessExit;
+
+            return Task.CompletedTask;
+        }
+
+        private void OnApplicationStarted()
+        {
+            Logger.LogInformation("Application started. Hosting environment: {envName}; Content root path: {contentRoot}",
+                Environment.EnvironmentName, Environment.ContentRootPath);
+
+            SystemdNotifier.Notify(ServiceState.Ready);
```

Comment:
It would seem reasonable, if `Notify` was `async`, to fire-and-forget here (maybe with some code to capture and log errors?). Since `OnApplicationStarted` is synchronous, we don't really have other options here.